### PR TITLE
Np 44887 new field external url

### PR DESCRIPTION
**Pull request by torarnet**

### CODE REVIEW

File: cristin-project/src/main/java/no/unit/nva/cristin/projects/model/nva/NvaProject.java [hunk 1]
```diff
@@ -373,6 +376,14 @@ public void setCreator(NvaContributor creator) {
         this.creator = creator;
     }
 
+    public URI getWebPage() {
```
Comment (brinxmat) [inline cristin-project/src/main/java/no/unit/nva/cristin/projects/model/nva/NvaProject.java:379]: This may be null, sometimes I wonder if Optional is a good return type to make this clear…possibly not in this case.

File: cristin-project/src/main/java/no/unit/nva/cristin/projects/model/nva/NvaProjectBuilder.java [hunk 2]
```diff
@@ -150,9 +151,14 @@ public NvaProject build() {
                    .withApprovals(extractApprovals(cristinProject.getApprovals()))
                    .withExemptFromPublicDisclosure(cristinProject.getExemptFromPublicDisclosure())
                    .withCreator(extractCreator(cristinProject.getCreator()))
+                   .withWebPage(extractWebPage(cristinProject.getExternalUrl()))
                    .build();
     }
 
+    private URI extractWebPage(String externalUrl) {
```
Comment (brinxmat) [inline cristin-project/src/main/java/no/unit/nva/cristin/projects/model/nva/NvaProjectBuilder.java:158]: Right the optional of getter does make sense

File: cristin-project/src/test/java/no/unit/nva/cristin/projects/fetch/project.feature [hunk 3]
```diff
@@ -167,3 +167,11 @@ Feature: API tests for Cristin Project retrieve and search
     And match response.approvals[0].applicationCode == 'EthicalApproval'
     And match response.approvals[0].identifier == '2017/800'
     And match response.approvals[0].authorityName == '#present'
+
+  Scenario: Fetch returns project with webpage
```
Comment (brinxmat) [inline cristin-project/src/test/java/no/unit/nva/cristin/projects/fetch/project.feature:171]: You might want the non-webpage test too

**Review by anetteOlli — DISMISSED** [review]
Lest igjennom. Fant ikke noe åpenbart å pirke på

File: cristin-project/src/main/java/no/unit/nva/cristin/projects/update/ProjectPatchValidator.java [hunk 4]
```diff
@@ -237,4 +240,15 @@ private void validateExtraLanguages(ObjectNode input) throws BadRequestException
     private BadRequestException notAListOfMapsException() {
         return new BadRequestException(format(NOT_A_VALID_LIST_OF_KEY_VALUE_FIELDS, ALTERNATIVE_TITLES));
     }
+
+    private void validateWebPage(ObjectNode input) throws BadRequestException {
+        if (input.has(WEB_PAGE)) {
```
Comment (StigNorland) [inline cristin-project/src/main/java/no/unit/nva/cristin/projects/update/ProjectPatchValidator.java:245]:
```
        Optional.ofNullable(input.get(WEB_PAGE))
            .map(JsonNode::asText)
            .map(URI::create)
            .orElseThrow(failure -> fieldIsNotUri());
```
If you want to follow the same style as in `CristinProjectBuilder.java`

File: cristin-project/src/test/java/no/unit/nva/cristin/projects/RandomProjectDataGenerator.java [hunk 5]
```diff
@@ -120,6 +121,7 @@ public static NvaProject randomNvaProject() {
                                           .withHealthProjectData(randomHealthProjectData())
                                           .withApprovals(randomApprovals())
                                           .withExemptFromPublicDisclosure(randomBoolean())
+                                          .withWebPage(randomUri())
```
Comment (StigNorland) [inline cristin-project/src/test/java/no/unit/nva/cristin/projects/RandomProjectDataGenerator.java:124]: How about adding a randomUri(withEndpoint) to give us a clue to where a randomUri should have pointed to?

**Review by StigNorland — APPROVED** [review]
I added some thoughts that has nothing to do with the code in question, except for use of Optional, which you may concider changing (adding)

File: cristin-project/src/main/java/no/unit/nva/cristin/projects/model/nva/NvaProject.java [hunk 6]
```diff
@@ -105,6 +105,9 @@ public class NvaProject implements JsonSerializable {
     @JsonProperty
     @JsonInclude(NON_NULL)
     private NvaContributor creator;
+    @JsonProperty
+    @JsonInclude(NON_NULL)
```
[the remark this answers is not in the document]
Comment (torarnet) [reply cristin-project/src/main/java/no/unit/nva/cristin/projects/model/nva/NvaProject.java:109]: So what you are saying is if a property does not have a value, we should display it as null in the json? Then i think we should use JsonInclude(ALWAYS) and not hide fields with value null?

[the remark this answers is not in the document]
Comment (StigNorland) [reply cristin-project/src/main/java/no/unit/nva/cristin/projects/model/nva/NvaProject.java:109]: That is my opinion (and it is easier to handle for the frontend).

**Review by ketilaa — APPROVED** [review]

File: cristin-project/src/test/java/no/unit/nva/cristin/projects/RandomProjectDataGenerator.java (hunk 5 above)
Comment (torarnet) [reply cristin-project/src/test/java/no/unit/nva/cristin/projects/RandomProjectDataGenerator.java:124]: this webpage can be anything, it does not have a common domain

File: cristin-project/src/main/java/no/unit/nva/cristin/projects/update/ProjectPatchValidator.java (hunk 4 above)
Comment (torarnet) [reply cristin-project/src/main/java/no/unit/nva/cristin/projects/update/ProjectPatchValidator.java:245]: input.has and input.get works a little different. Having something to do with the field present or not, and the field being null or not when present. Tests failed when i tried you example

File: cristin-project/src/main/java/no/unit/nva/cristin/projects/model/nva/NvaProjectBuilder.java (hunk 2 above)
Comment (torarnet) [reply cristin-project/src/main/java/no/unit/nva/cristin/projects/model/nva/NvaProjectBuilder.java:158]: no other field in this model class is using optional. if we want to start using it i suggest we do it for all optional fields in a separate PR. we should also discuss one day using field = null or field omitted if there is no value there.

File: cristin-project/src/main/java/no/unit/nva/cristin/projects/update/ProjectPatchValidator.java (hunk 4 above)
Comment (StigNorland) [reply cristin-project/src/main/java/no/unit/nva/cristin/projects/update/ProjectPatchValidator.java:245]: Did you know that ?
Wouldn't that have implications for when we are serializing empty fields to storage.
(we omit them, instead of adding with null)

File: cristin-project/src/test/java/no/unit/nva/cristin/projects/fetch/project.feature (hunk 3 above)
Comment (torarnet) [reply cristin-project/src/test/java/no/unit/nva/cristin/projects/fetch/project.feature:171]: We might need to decide if field should be null or just omitted first, but i can always add a test now and change it later

File: cristin-project/src/main/java/no/unit/nva/cristin/projects/update/ProjectPatchValidator.java (hunk 4 above)
Comment (torarnet) [reply cristin-project/src/main/java/no/unit/nva/cristin/projects/update/ProjectPatchValidator.java:245]: yes i did not it but tested your suggestion just to see if we could simplify. When doing patch requests we have to build the json manually because patch allows us to explicit send a field value as null, so we have to know if the client sends the field at all, sends the field with value null, or field with a regular value

Comment (torarnet) [reply cristin-project/src/main/java/no/unit/nva/cristin/projects/update/ProjectPatchValidator.java:245]: Also added a test for checking if field will be sent as null to upstream

Comment (StigNorland) [reply cristin-project/src/main/java/no/unit/nva/cristin/projects/update/ProjectPatchValidator.java:245]: Maybe something that the whole team should know about ?

**Review by StigNorland — APPROVED** [review]

**Review by brinxmat — APPROVED** [review]
